[docs] update design.md doc (#6359)
open-telemetry/opentelemetry-collector: docs/design.md

# OpenTelemetry Collector Architecture

This document describes the architecture design and implementation of the
OpenTelemetry Collector.

## Summary

OpenTelemetry Collector is an executable that can receive telemetry data, optionally process it, and export it further.

The Collector supports several popular open-source protocols for receiving and sending telemetry data as well as offering a pluggable architecture for adding more protocols.

Data receiving, processing, and exporting is done using [Pipelines](#pipelines). The Collector can be configured to have one or more pipelines. Each pipeline includes:

- a set of [Receivers](#receivers) that receive the data
- a series of optional [Processors](#processors) that get the data from receivers and process it
- a set of [Exporters](#exporters) which get the data from processors and send it further outside the Collector.

> The same receiver can feed data to multiple pipelines and multiple pipelines can feed data into the same Exporter.

Reviewer: ```suggestion
The same receiver can be included in multiple pipelines and multiple pipelines can include the same Exporter.
```

This read as if receivers and exporters were separate from pipelines, which I don't think is the intent.

## Pipelines

A pipeline defines a path the data follows in the Collector starting from reception, then further processing or modification and finally exiting the Collector via exporters.

Pipelines can operate on 3 telemetry data types: traces, metrics, and logs. The data type is a property of the pipeline defined by its configuration. Receivers, processors, and exporters used in a pipeline must support the particular data type otherwise `ErrDataTypeIsNotSupported` will be reported when the configuration is loaded. A pipeline can be depicted the following way:

![Pipelines](images/design-pipelines.png)

There can be one or more receivers in a pipeline. Data from all receivers is pushed to the first processor, which performs a processing on it and then pushes it to the next processor (or it may drop the data, e.g. if it is a “sampling” processor) and so on until the last processor in the pipeline pushes the data to the exporters. Each exporter gets a copy of each data element. The last processor uses a `fanoutconsumer` to fan out the data to multiple exporters.

> The pipeline is constructed during Collector startup based on pipeline definition in the config file.

Reviewer: ```suggestion
The pipeline is constructed during Collector startup based on pipeline definition in the configuration.
```

Might be good to avoid referring to config "files" as there are now multiple sources of configuration.

```yaml
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: localhost:4317

service:
  pipelines:
    traces:  # a pipeline of “traces” type
      receivers: [otlp]
      processors: [memory_limiter, batch]
      exporters: [jaeger]
    traces/2:  # another pipeline of “traces” type
      receivers: [otlp]
      processors: [batch]
      exporters: [opencensus]
```

> In the above example `otlp` receiver will send the same data to pipeline `traces` and to pipeline `traces/2`. (Note: the configuration uses composite key names in the form of `type[/name]` as defined in [this document](https://docs.google.com/document/d/1NeheFG7DmcUYo_h2vLtNRlia9x5wOJMlV4QKEK05FhQ/edit#)).

Reviewer: Can we get `this document` included in this repo and not on a gdoc of unknown provenance?

Author: See PR here https://github.com/open-telemetry/opentelemetry-collector/pull/6369

When the Collector loads this config the result will look like this (part of processors and exporters are omitted from the diagram for brevity):

![Receivers](images/design-receivers.png)

Important: when the same receiver is referenced in more than one pipeline the Collector will create only one receiver instance at runtime that will send the data to a fan out consumer, which in turn will send the data to the first processor of each pipeline. The data propagation from receiver to the fan out consumer and then to processors is done via a synchronous function call. This means that if one processor blocks the call the other pipelines that are attached to this receiver will be blocked from receiving the same data and the receiver itself will stop processing and forwarding newly received data.

### Exporters

> Exporters typically forward the data they get to a destination on a network (but they can also send it elsewhere, e.g `logging` exporter writes the telemetry data to a standard out).

Reviewer: ```suggestion
Exporters typically forward the data they get to a destination on a network (but they can also send it elsewhere, e.g `logging` exporter writes the telemetry data to the logging destination).
```

Not sure about the wording here, but IIRC the `logging` exporter will wrap the service logger, which may be configured to write somewhere other than `STDOUT`.

A pipeline can contain sequentially connected processors. The first processor gets the data from one or more receivers that are configured for the pipeline, the last processor sends the data to one or more exporters that are configured for the pipeline. All processors between the first and last receive the data strictly only from one preceding processor and send data strictly only to the succeeding processor.

Processors can transform the data before forwarding it (i.e. add or remove attributes from spans), they can drop the data simply by deciding not to forward it (this is for example how the `probabilisticsampler` processor works), they can also generate new data. This is how a `spanmetrics` processor can produce metrics for spans processed by the pipeline.

> The same name of the processor can be referenced in the `processors` key of multiple pipelines. In this case the same configuration will be used for each of these processors however each pipeline will always get its own instance of the processor. Each of these processors will have its own state, the processors are never shared between pipelines. For example if `batch` processor is used in several pipelines each pipeline will have its own batch processor (although the batch processor will be configured exactly the same way if the reference the same key in the config file). As an example, given the following config:

Reviewer: ```suggestion
The same name of the processor can be referenced in the `processors` key of multiple pipelines. In this case the same configuration will be used for each of these processors however each pipeline will always get its own instance of the processor. Each of these processors will have its own state, the processors are never shared between pipelines. For example if `batch` processor is used in several pipelines each pipeline will have its own batch processor (although each batch processor will be configured exactly the same way if they reference the same key in the configuration). As an example, given the following configuration:
```

To resolve the issues above, you can run OpenTelemetry Collector as an Agent.
The Agent runs as a daemon in the VM/container and can be deployed independent
of Library. Once Agent is deployed and running, it should be able to retrieve

> spans/stats/metrics from Library, export them to other backends. We MAY also

Reviewer 2: s/spans/traces?

Reviewer: I think "spans" is correct.  The data the collector operates on are individual spans.  The data model doesn't even have a "trace" entity, just a collection of "spans".

Author: Right, but the configuration for pipelines references `traces`, which aligns with the signal terminology for spans. I guess this is a bit of a grey area :(

Reviewer 2: The reason to use "traces" is because that's how we name the pipeline type.

> spans/metrics/logs from Library, export them to other backends. We MAY also

Author: ```suggestion
traces/metrics/logs from Library, export them to other backends. We MAY also
```

![agent-architecture](images/design-collector-agent.png)

For developers/maintainers of other libraries: Agent can also

> accept spans/metrics/logs from other tracing/monitoring libraries, such as

Author: ```suggestion
accept traces/metrics/logs from other tracing/monitoring libraries, such as
```
